crates/pallet-executor/src/lib.rs

```diff
@@ -430,6 +511,21 @@ impl<T: Config> Pallet<T> {
             return Err(Error::<T>::UnexpectedExecutionReceiptSigner);
         }
 
+        // Ensure the receipt is neither old nor too new.
+        let primary_number = execution_receipt.primary_number;
+
+        let best_number = ExecutionChainBestNumber::<T>::get();
+        if primary_number <= best_number {
+            return Err(Error::<T>::ExecutionReceiptStale);
+        }
+
+        let current_block_number = frame_system::Pallet::<T>::current_block_number();
+        if primary_number == current_block_number
+            || primary_number > best_number + T::MaximumReceiptDrift::get()
```

Comment:
Shouldn't it just be this?:
```suggestion
        if primary_number > current_block_number
```

I think `current_block_number` is the previous block, so whenever we receive ER in the very next block `primary_number == current_block_number` will render ER invalid and cause it to be accepted with a delay in the next block after this.

And I do not understand why you need a drift here that might run past `current_block_number`, which in turn should be physically impossible and thus invalid.

Reply:
Okay, a more descriptive name might be helpful here. `current_block_number` is the current consensus chain number, `best_number` is the best execution chain number. The execution chain can lag behind the consensus chain due to no ERs being submitted in time.

- `primary_number == current_block_number` is because the receipt must point to a previous block, you can't submit a receipt for a block that is being built right now.
- `primary_number > best_number + T::MaximumReceiptDrift::get()` means we only accept a reason number of future ERs that are ahead of the current best execution chain number. The new receipts should extend the execution chain instead of the consensus chain, so we compare it against the execution chain best number.

Reply:
> current_block_number is the current consensus chain number

I do not think so. It is the tip of the consensus chain. Validation is called by transaction pool to validate transactions and I don't think it is called within block production from within runtime, rather transaction pools calls here to check whether transaction became valid after new block was produced.

> means we only accept a reason number of future ERs that are ahead of the current best execution chain number

But why? `best_number + T::MaximumReceiptDrift::get()` can easily be higher than `current_block_number`, which doesn't make sense.

The check I suggested seems to cover both of these cases in a simpler to understand way.

Reply:
>I do not think so. It is the tip of the consensus chain. Validation is called by transaction pool to validate transactions and I don't think it is called within block production from within runtime, rather transaction pools calls here to check whether transaction became valid after new block was produced.

Yeah, I meat the tip of the chain, but said in a confusing way. Validation is not called within block production but it is called against the tip of the chain to see if they are ready to be included in the next block, for example, the context for performing the validation is the latest imported number + 1(the next block number) when you submit an extrinsic.

>But why? best_number + T::MaximumReceiptDrift::get() can easily be higher than current_block_number, which doesn't make sense.

There is a case that the local height is lower than the network height due to the network delay, but somehow it received the receipts for the valid futures blocks so we cache a number of these receipts in the transaction pool, no? Simply using `if primary_number > current_block_number` will just reject these kind of receipts unconditionally.

Reply:
> Yeah, I meat the tip of the chain, but said in a confusing way. Validation is not called within block production but it is called against the tip of the chain to see if they are ready to be included in the next block, for example, the context for performing the validation is the latest imported number + 1(the next block number) when you submit an extrinsic.

I had to test it, but you're right, I had an impression that it'll test against current block number. `primary_number >= current_block_number` then?

> There is a case that the local height is lower than the network height due to the network delay, but somehow it received the receipts for the valid futures blocks so we cache a number of these receipts in the transaction pool, no? Simply using if primary_number > current_block_number will just reject these kind of receipts unconditionally.

From what I understand it will consider them invalid for now, but then transaction might become valid later, no? If so, then additional concept of `MaximumReceiptDrift` is unnecessary.

Reply:
>primary_number >= current_block_number then?

Unless we clearly reject any potential valid future receipts(`MaximumReceiptDrift`), `primary_number >= current_block_number` will cause conflict with `primary_number > best_number + T::MaximumReceiptDrift::get()`.

>From what I understand it will consider them invalid for now, but then transaction might become valid later, no? If so, then additional concept of MaximumReceiptDrift is unnecessary.

It's not considered as invalid, but will put into the futures queue which requires some prerequisites to be satisfied, see the test case added in https://github.com/subspace/subspace/pull/436/commits/11dbaf85817f2c9ee780a12d0874c70e34e8012e. From what I understand, this is the out-of-order issue caused by the network delay as we discussed before, unless I have some misunderstanding here.

Reply:
OK, I understand what you said now and satisfied with the answer, thanks!